File: addons/payment_custom/static/src/js/post_processing.js
/** @odoo-module **/

import paymentPostProcessing from '@payment/js/post_processing';

paymentPostProcessing.include({
    /**
     * Don't wait for the transaction to be confirmed before redirecting customers to the
     * landing route because custom transactions remain in the state 'pending' forever.
     *
     * @override method from `@payment/js/post_processing`
     * @param {string} providerCode - The code of the provider handling the transaction.
     */
    _getFinalStates(providerCode) {
        const finalStates = this._super(...arguments);
        if (providerCode === 'custom') {
            finalStates.add('pending');
        }
        return finalStates;
    }
});

File: addons/pos_online_payment/controllers/post_processing.py
from odoo.addons.payment.controllers.post_processing import PaymentPostProcessing


class PosPaymentPostProcessing(PaymentPostProcessing):

    def get_payment_status_template_xmlid(self, tx):
        if tx and tx.pos_order_id:
            return 'pos_online_payment.pos_payment_status'
        return super().get_payment_status_template_xmlid(tx)
